src/Microsoft.ML.AutoML/DatasetDimensions/ReservoirSampledDataView.cs

```diff
@@ -0,0 +1,315 @@
+        }
+
+        public DataViewRowCursor GetRowCursor(IEnumerable<DataViewSchema.Column> columnsNeeded, Random rand = null)
+        {
+            if (!_initialized)
+            {
+                Initialize();
+            }
+            return new Cursor(Schema, _sampledColumns, GetRowCount().Value);
+        }
+
+        public DataViewRowCursor[] GetRowCursorSet(IEnumerable<DataViewSchema.Column> columnsNeeded, int n, Random rand = null)
+        {
+            if (!_initialized)
+            {
+                Initialize();
+            }
+            return new DataViewRowCursor[] { new Cursor(Schema, _sampledColumns, GetRowCount().Value) };
+        }
+
+        /// <summary>
+        /// Initializes the data view.
+        /// Iterates through every row to build the reservoir sample.
+        /// </summary>
+        private void Initialize()
+        {
+            var cursor = _originalData.GetRowCursor(Schema);
+
+            // Init data structures for each column
+            var columnSampleBuilders = new ColumnSampleBuilder[Schema.Count];
+            for (var i = 0; i < Schema.Count; i++)
+            {
+                // Ignore inactive columns
+                if (!cursor.IsColumnActive(Schema[i]))
+                {
+                    continue;
+                }
+
+                var colType = Schema[i].Type;
+
+                // Init SampledColumn
+                Type sampledColumnType = colType.IsVector() ?
+                    typeof(VectorSampledColumn<>).MakeGenericType(colType.GetItemType().RawType) :
+                    typeof(SingleSlotSampledColumn<>).MakeGenericType(colType.RawType);
+                var constructor = sampledColumnType.GetConstructor(new[] { typeof(int) });
+                _sampledColumns[i] = constructor.Invoke(new object[] { _sampleSize });
+
+                // Init ColumnSampleBuilder
+                var columnSampleBuilderType = typeof(ColumnSampleBuilder<>).MakeGenericType(colType.RawType);
+                var originalColumnGetter = MLNetUtils.MarshalInvoke(GetGetter<int>, colType.RawType, cursor, i);
+                constructor = columnSampleBuilderType.GetConstructor(new[] { sampledColumnType, originalColumnGetter.GetType() });
+                columnSampleBuilders[i] = constructor.Invoke(new object[] { _sampledColumns[i], originalColumnGetter }) as ColumnSampleBuilder;
+            }
+
+            var sampleBuilder = new SampleBuilder(columnSampleBuilders);
+
+            // Reservoir sample over every row
+            long rowIdx = 0;
+            while (cursor.MoveNext())
+            {
+                // Cache first _sampleSize rows
+                if (rowIdx < _sampleSize)
+                {
+                    sampleBuilder.StoreNextRow(rowIdx);
+                }
+                else
+                {
+                    // Reservoir sample row
+                    var idx = GetRandomLong(rowIdx + 1);
```

Comment:
I suspect a multiply is a faster than the current function w/ a mod...
```suggestion
                    var idx = rand.NextDouble() * (rowIdx + 1);
```

Reply:
You're right lol -- I timed this out
Generating 10 million random #s w/ the rand.NextDouble() took 228 milliseconds
Generating 10 million random #s w/ the existing method took 1081 milliseconds
Both are quite fast, which is good.
This is probably a theoretical concern instead of an actual issue -- but b/c of the fixed double precision, I find myself wondering if rand.NextDouble() * a long that is sufficiently large could produce a less than random distribution. For instance, in the extreme, I observe that 1.0 / (long.MaxValue/100 - 5) == 1.0 / (long.MaxValue/100 - 4) evaluates to true. Clearly, we would never have anywhere NEAR long.MaxValue rows, but this is just to illustrate the point at the extreme. I have a strong hunch this is more of a theoretical concern than a practical one, but just throwing it out there